File: src/main/swift/g0701_0800/s0739_daily_temperatures/Solution.swift
// #Medium #Top_100_Liked_Questions #Array #Stack #Monotonic_Stack #Programming_Skills_II_Day_6
// #Big_O_Time_O(n)_Space_O(n) #2024_07_03_Time_535_ms_(99.24%)_Space_26.9_MB_(19.32%)

class Solution {
    func dailyTemperatures(_ temperatures: [Int]) -> [Int] {
        var sol = [Int](repeating: 0, count: temperatures.count)
        sol[temperatures.count - 1] = 0
        for i in (0..<sol.count - 1).reversed() {
            var j = i + 1
            while j < sol.count {
                if temperatures[i] < temperatures[j] {
                    sol[i] = j - i
                    break
                } else {
                    if sol[j] == 0 {
                        break
                    }
                    j += sol[j]
                }
            }
        }
        return sol
    }
}

File: src/main/c/g0701_0800/s0739_daily_temperatures/Solution.c
// #Medium #Top_100_Liked_Questions #Array #Stack #Monotonic_Stack #Programming_Skills_II_Day_6
// #Big_O_Time_O(n)_Space_O(n) #2024_11_08_Time_3_ms_(98.31%)_Space_68.5_MB_(94.81%)

#include <stdio.h>
#include <stdlib.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
// Function to calculate the daily temperatures
int* dailyTemperatures(int* temperatures, int temperaturesSize, int* returnSize) {
    int* sol = (int*)calloc(temperaturesSize, sizeof(int)); // Initialize the solution array with zeros
    sol[temperaturesSize - 1] = 0; // Last element is 0 by default
    *returnSize = temperaturesSize; // Set the size of the return array

    for (int i = temperaturesSize - 2; i >= 0; i--) {
        int j = i + 1;
        while (j < temperaturesSize) {
            if (temperatures[i] < temperatures[j]) {
                sol[i] = j - i;
                break;
            } else {
                if (sol[j] == 0) {
                    break;
                }
                j = j + sol[j];
            }
        }
    }

    return sol;
}
